vowpalwabbit/global_data.h

```diff
@@ -359,6 +359,342 @@ struct vw_logger
+
+  uint32_t holdout_period;
+  uint32_t holdout_after;
+};
+
+struct InputConfig
+{
+  bool stdin_off;
+
+  /*
+   * Set once on; on parse_args
+   * Input file
+   */
+  std::string data_filename;  // was vm["data"]
+
+  /*
+   * controversial, weight related
+   * but this is a list of file names to be read
+   */
+  std::vector<std::string> initial_regressors;
+
+  /*
+   * input related, since its a file name to be read
+   * might be mostly gd.cc related, can used for initial weights
+   */
+  std::string feature_mask;
+
+  /*
+   * input related, bc filename
+   * might be mostly bfgs.cc related?
+   * related to weights
+   */
+  std::string per_feature_regularizer_input;
+
+  /* input, example related
+   * feature dictionary stuff, applies to the next 3
+   */
+  std::vector<std::string> dictionary_path;            // where to look for dictionaries
+};
+
+struct OutputConfig
+{
+  bool save_per_pass;
+
+  /*
+   * output related config
+   */
+  bool save_resume;
+
+  /*
+   * output related config
+   */
+  std::string id;
+
+  /*
+   * GD specific config, but not really
+   */
+  bool preserve_performance_counters;
+
+  /*
+   * output configuration
+   */
+  // error reporting
+  vw_ostream trace_message;
+
+  /*
+   * output related
+   */
+  std::unique_ptr<VW::io::writer> stdout_adapter;
+
+  /*
+   * output related
+   * might be mostly bfgs.cc related?
+   * applies to both
+   */
+  std::string per_feature_regularizer_output;
+  std::string per_feature_regularizer_text;
+
+  /* flag for runtime behaviour, input setting, for output
+   */
+  bool audit;     // should I print lots of debugging information?
+
+  std::string final_regressor_name;
+  std::string text_regressor_name;
+  std::string inv_hash_regressor_name;
+
+  bool hash_inv;
+
+  /*this could be state, but we can clean up impl
+   * def a code smell, we keep passing them along */
+  // Set by --progress <arg>
+  bool progress_add;   // additive (rather than multiplicative) progress dumps
+  float progress_arg;  // next update progress dump multiplier
+
+};
+
+struct VWRuntimeConfig
+{
+  /*
+   * Set once on; on parse_args
+   * runtime configuration
+   * also applies to num_children
+   */
+  bool daemon;
+  size_t num_children;
+
+  /*
+   * runtime configuration
+   */
+  bool vw_is_main = false;  // true if vw is executable; false in library mode
+
+  /*
+   * runtime configuration
+   * not set directly by user
+   */
+  // Flag used when VW internally manages lifetime of options object.
+  bool should_delete_options = false;
+
+  uint64_t random_seed;
+
+  bool no_daemon = false;  // If a model was saved in daemon or active learning mode, force it to accept local input when loaded instead.
+
+};
+
+struct GlobalState
+{
+  time_t init_time;
+
+  /*
+   * Should be part of learner?
+   * not input, used as state
+   */
+  uint64_t current_pass;
+
+  /*
+   * not input, used as state
+   * related to examples
+   */
+  size_t passes_complete;
```

Comment:
I _think_ `current_pass` and `passes_complete` are almost the same thing. AFAIK the only difference is that `passes_complete` counts based on the input file (incremented once an input is fully read and reset), and `current_pass` is based on the example itself (incremented after the last example in a file is processed).

Maybe add a todo to ensure these two variables are being used in a consistent way, If my understanding is correct, `passes_complete` should only ever be used in the parser thread while `current_pass` should only ever be used in the learner thread